drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -1025,114 +1025,86 @@ impl Transceivers {
-            .fpga(FpgaController::Left)
-            .write_bytes(WriteOp::Write, Addr::QSFP_WRITE_BUFFER, buf)
-            .is_ok()
-        {
-            success |= LEFT_LOGICAL_MASK;
-        }
-        if self
-            .fpga(FpgaController::Right)
-            .write_bytes(WriteOp::Write, Addr::QSFP_WRITE_BUFFER, buf)
-            .is_ok()
-        {
-            success |= RIGHT_LOGICAL_MASK;
-        }
-        let error = !success;
-
-        ModuleResultNoFailure::new(success, error).unwrap_lite()
-    }
+    pub fn set_i2c_write_buffer(
+        &self,
+        buf: &[u8],
+        mask: LogicalPortMask,
+    ) -> ModuleResultNoFailure {
+        let mut error = LogicalPortMask(0);
 
-    /// For a given `local_port`, return the Addr where its read buffer begins
-    pub fn read_buffer_address(local_port: PhysicalPort) -> Addr {
-        match local_port.0 % 16 {
-            0 => Addr::QSFP_PORT0_READ_BUFFER,
-            1 => Addr::QSFP_PORT1_READ_BUFFER,
-            2 => Addr::QSFP_PORT2_READ_BUFFER,
-            3 => Addr::QSFP_PORT3_READ_BUFFER,
-            4 => Addr::QSFP_PORT4_READ_BUFFER,
-            5 => Addr::QSFP_PORT5_READ_BUFFER,
-            6 => Addr::QSFP_PORT6_READ_BUFFER,
-            7 => Addr::QSFP_PORT7_READ_BUFFER,
-            8 => Addr::QSFP_PORT8_READ_BUFFER,
-            9 => Addr::QSFP_PORT9_READ_BUFFER,
-            10 => Addr::QSFP_PORT10_READ_BUFFER,
-            11 => Addr::QSFP_PORT11_READ_BUFFER,
-            12 => Addr::QSFP_PORT12_READ_BUFFER,
-            13 => Addr::QSFP_PORT13_READ_BUFFER,
-            14 => Addr::QSFP_PORT14_READ_BUFFER,
-            15 => Addr::QSFP_PORT15_READ_BUFFER,
-            _ => unreachable!(),
+        // for every port in the mask
+        for port in mask.to_indices() {
+            // we need to know the FPGA and which physical port
+            let loc = port.get_physical_location();
+            let fpga = self.fpga(loc.controller);
+            let port_addr = loc.port.0 as u16;
+            // clear the FIFO
+            if fpga
+                .write(
+                    WriteOp::BitSet,
+                    Addr::QSFP_PORT0_CONTROL as u16 + port_addr,
+                    Reg::QSFP::PORT0_CONTROL::WDATA_FIFO_CLEAR,
+                )
+                .is_ok()
+            {
+                // write in teh buffer
+                if fpga
+                    .write_bytes(
+                        WriteOp::WriteNoAddrIncr,
+                        Addr::QSFP_PORT0_I2C_DATA as u16 + port_addr,
+                        buf,
+                    )
+                    .is_err()
+                {
+                    error |= port.as_mask();
+                }
+            } else {
+                error |= port.as_mask();
+            }
         }
-    }
+        // success is wherever we did not encounter an `FpgaError`
+        let success = mask & !error;
+        // only have an error where there was a requested module in mask
+        error &= mask;
```

Comment:
I'm not totally sure if I get this --- IIUC, since we only write to the FPGA for ports in the mask, we shouldn't have to mask out the error bits again, right? Or am I missing something?

Reply:
You have not missed anything and are correct! This was a bit of copy/pasta on my part when reworking some similar code where the initial `error` mask could not be constructed intelligently (by port) and therefore I masked it off later. I'll clean this up!